### PR TITLE
Backport to 1.7.10

### PR DESCRIPTION
**Pull request by DarkShadow44**
This is a backport of the complete mod (all features working) to 1.7.10.

A few major changes:
- Most ASM is handles by the great Sponge Mixin library
- Dependency on AppleCore for crop growth events
- Per biome config is disabled since we don't have useful ids (Tropical seasons are hardcoded to deserts)

I made sure it's compatible with most other mods that are affected by seasons, for example BiomesOPlenty, BetterFoliage and DynamicSourroundings.

If something should be broken or missing, of course I will be available for a bugfix.

### CODE REVIEW

**Comment by Adubbz** [comment]
Thank you! I haven't gotten a chance to review this properly yet, although I see you've noted that there's a dependency on AppleCore. I'd prefer if SS remains a standalone mod, as dependencies are always something that I've not been too keen on.

**Comment by DarkShadow44** [comment]
I understand the desire for a standalone mod, but crop growing events don't exist in 1.7.10 yet.  I could make AppleCore an optional dependency though, if that would help?
The other option would be to make the hooks myself, which would be problematic for AppleCore compatibility.

**Comment by Adubbz** [comment]
Ideally the hooks would be implemented in SS, yeah. If it’s possible to do so in a way that’s compatible with AppleCore that’d be the best option. If not, let me know

**Comment by DarkShadow44** [comment]
So, the point is to remain standalone, right? Would it be okay to copy those hooks from AppleCore (the license permits that) and just use them? That would reduce the amount of work drastically and should be compatible, while allowing SS to be standalone.

**Comment by Adubbz** [comment]
You'd need to check if their license allows for that, but if it does, then sure, go for it.

**Comment by DarkShadow44** [comment]
I added a new commit that should solve the problem. Applying the hooks twice is too tricky, so here's what I changed:
When AppleCore is present, we use it's API - this is the only reasonable way to work around compatibility issues, although we need to manually detect AppleCore (see the changes in SSLoadingPlugin).
When AppleCore is not present, we use a small portion of its code (Unlicensed, so that's legal) to add crop functionality - so we don't have a hard dependency without losing any features.

How does that look?

**Comment by Adubbz** [comment]
That works for me, perfect. Another general comment: Could you replace all your tabs with 4 spaces? I realise there are some inconsistencies that have slipped in over time in the main branch, although the intent is to use spaces.

**Comment by DarkShadow44** [comment]
Sure thing, I add another commit to fix the formatting. If you want me to squash the changes into the current commits, just tell me.

File: src/main/java/sereneseasons/mixins/BlockMixin.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+     * @reason Redirect to our coloring handlers.
+     */
+    @Overwrite
+    public int colorMultiplier(IBlockAccess blockAccess, int x, int y, int z)
+    {
+        Block block = (Block) (Object) this;
+
+        if (block instanceof BlockGrass)
+        {
+            return handleGras(blockAccess, x, y, z);
```
Comment (Adubbz) [inline src/main/java/sereneseasons/mixins/BlockMixin.java]: handleGrass

**Comment by Adubbz** [comment]
Aside from that minor typo above, I think I'm pretty happy with the code actually. Will merge once that's fixed.

**Comment by DarkShadow44** [comment]
Thanks, I pushed a fix.

**Comment by Mnesikos** [comment]
Hey, @DarkShadow44! I'm part of a 1.7.10 server that just tried adding your backport today, but couldn't get it up and running. Any chance you can take an extra look to have this working server side, particularly with Cauldron-1.7.10-2.1403.1.54?

**Comment by DarkShadow44** [comment]
What does "couldn't get it up and running" mean exactly? Do you have logs?

**Comment by Mnesikos** [comment]
Sorry for the lack of detail there. I have this much from the server owner: 
`Mixin apply failed mixins.seasons.json:BlockMixin -> net.minecraft.block.Block: org.spongepowered.asm.mixin.transformer.throwables.InvalidMixinException Overwrite target "func_149720_d" was not located in target class net/minecraft/block/Block
org.spongepowered.asm.mixin.transformer.throwables.InvalidMixinException: Overwrite target "func_149720_d" was not located in target class net/minecraft/block/Block`

I will ask for a full log from them and share that if possible. However, I also cloned your repo and tried booting it up myself server side, but received this crash: [https://hastebin.com/eqerowuyab.sql](https://hastebin.com/eqerowuyab.sql)

**Comment by DarkShadow44** [comment]
@Mnesikos I'll track the issue here: https://github.com/DarkShadow44/SereneSeasons/issues/1
I'll fix that.

**Comment by danielrossyamitrano** [comment]
I have little hope that this issue will be fixed, but I my game crashed when reloading the world because the Seasons Clock bounced and then crashed. [Here's the crash report](https://pastebin.com/mumQnH5n).

**Comment by DarkShadow44** [comment]
Please take any issues to https://github.com/DarkShadow44/SereneSeasons/issues
But first make sure to build my latest code from source, it might already be fixed. If not, I'll fix it.

**Comment by danielrossyamitrano** [comment]
@DarkShadow44 I already have what I think is the lastest version (1.2.18)

and also, there is already an open issue there with [the same topic](https://github.com/DarkShadow44/SereneSeasons/issues/20).

**Pull request merged**